src/main/scala/eu/proteus/solma/sampling/AdaptiveReservoirSampling.scala

```diff
@@ -0,0 +1,263 @@
+  def factorial(n: Int): Int = {   if (n == 0) 1 else n * factorial(n-1)  }
+  def min_m (i: Int,ReservoirSize: Int,delta: Int,threshold: Double) : Int = {
+
+    var m = delta
+    var UC = 0.0
+    var x = ReservoirSize
+    var f1 = factorial(i)*factorial(m)*factorial(ReservoirSize+delta)*factorial(i+m-ReservoirSize-delta)
+    var f2 = factorial(x)*factorial(i-x)*factorial(ReservoirSize+delta-x)*factorial(m-ReservoirSize-delta+x)*factorial(i+m)
+    UC = UC + f1/f2
+    while(UC <= threshold){
+      m += 1
+      UC = 0.0
+      for (x <- math.max(0,ReservoirSize+delta-m) to ReservoirSize){
+        f1 = factorial(i)*factorial(m)*factorial(ReservoirSize+delta)*factorial(i+m-ReservoirSize-delta)
+        f2 = factorial(x)*factorial(i-x)*factorial(ReservoirSize+delta-x)*factorial(m-ReservoirSize-delta+x)*factorial(i+m)
+        UC = UC + f1/f2
+      }
+    }
+    m
+  }
+
+  // ==================================== Factory methods ==========================================
+
+  def apply(): AdaptiveReservoirSampling = {
+    new AdaptiveReservoirSampling()
+  }
+
+
+  // ==================================== Operations ==========================================
+
+
+  implicit def fitNoOp[T] = {
+    new StreamFitOperation[AdaptiveReservoirSampling, T]{
+      override def fit(
+                        instance: AdaptiveReservoirSampling,
+                        fitParameters: ParameterMap,
+                        input: DataStream[T])
+      : Unit = {}
+    }
+  }
+
+
+  implicit def treansformAdaptiveReservoirSampling[T <: Vector : TypeInformation : ClassTag] = {
+    new TransformDataStreamOperation[AdaptiveReservoirSampling, T, mutable.ArrayBuffer[T]] {
+      override def transformDataStream(
+                                        instance: AdaptiveReservoirSampling,
+                                        transformParameters: ParameterMap,
+                                        input: DataStream[T])
+      : DataStream[mutable.ArrayBuffer[T]] = {
+        val resultingParameters = instance.parameters ++ transformParameters
+        val statefulStream = FlinkSolmaUtils.ensureKeyedStream[T](input, resultingParameters.get(PartitioningOperation))
+        var k = resultingParameters(AdaptiveReservoirSize)
+        val nboCh=resultingParameters(AdaptiveReservoirNumberToChange)
+        val thresh=resultingParameters(AdaptiveReservoirThreshold)
+        val changedReservoirSize = resultingParameters(Change)
+        val gen = new XORShiftRandom()
+        implicit val typeInfo = TypeInformation.of(classOf[(Long,Long,Long,mutable.ArrayBuffer[T],mutable.ArrayBuffer[T], mutable.ArrayBuffer[T])])
+        implicit val retInfo = TypeInformation.of(classOf[mutable.ArrayBuffer[T]])
+        var j : Int = 0
+        var m : Int = 0
+        var x : Int = 0
+        var remainingSize = 0
+        statefulStream.flatMapWithState((in, state: Option[(Long,Long,Long,mutable.ArrayBuffer[T],mutable.ArrayBuffer[T],mutable.ArrayBuffer[T])]) => {
```

Comment:
Too big tuples should be avoided. Perhaps, you may consider to use a case class.